sporkfish/searcher/negamax.py

```diff
@@ -0,0 +1,251 @@
+        Null move pruning - reduce the search space by trying a null move,
+        then seeing if the score of the subtree search is still high enough to cause a beta cutoff
+        """
+        # TODO: add zugzwang check
+        # Will make depth_reduction_factor configurable later
+        depth_reduction_factor = 3
+        in_check = board.is_check()
+        if depth >= depth_reduction_factor and not in_check:
+            null_move_depth = depth - depth_reduction_factor
+            board.push(chess.Move.null())
+            value = -self._negamax(board, null_move_depth, -beta, -alpha)
+            board.pop()
+            if value >= beta:
+                return True
+        return False
+
+    def _searcher(
+        self,
+        board: Board,
+        depth: int,
+        alpha: float,
+        beta: float,
+    ) -> Tuple[float, chess.Move]:
+        """
+        Entry point for negamax search with fail-soft alpha-beta pruning, single process.
+
+        :param board: The current chess board position.
+        :type board: Board
+        :param depth: The current search depth.
+        :type depth: int
+        :param alpha: Alpha value for alpha-beta pruning.
+        :type alpha: float
+        :param beta: Beta value for alpha-beta pruning.
+        :type beta: float
+        :return: Tuple containing the best move and its value.
+        :rtype: Tuple[float, chess.Move]
+        """
+        value = -float("inf")
+        best_move = chess.Move.null()
+        self._statistics.increment()
+
+        if self._config.enable_transposition_table:
+            hash_value = self._zobrist_hash.hash(board)
+
+        legal_moves = self._ordered_moves(board, board.legal_moves)
+        for move in legal_moves:
+            board.push(move)
+            child_value = -self._negamax(board, depth - 1, -beta, -alpha)
+            board.pop()
+
+            if value < child_value:
+                value = child_value
+                best_move = move
+
+            alpha = max(alpha, value)
+            if alpha >= beta:
+                break
+
+        if self._config.enable_transposition_table:
+            self._transposition_table.store(hash_value, depth, value)
+
+        return value, best_move
+
+    def search(
+        self, board: Board, timeout: Optional[float] = None
+    ) -> Tuple[float, chess.Move]:
+        """
+        Finds the best move (and associated score) via negamax and iterative deepening.
+
+        :param board: The current chess board position.
+        :type board: Board
+        :return: The best score and move based on the search.
+        :param timeout: Time in seconds until we stop the search, returning the best depth if we timeout.
+        :type timeout: Optional[float]
+        :rtype: Tuple[float, Move]
+        """
+
+        score, move = self._iterative_deepening(board, timeout)
+        return score, move
+
+
+# This doesn't really work yet. Don't use.
+class NegaMaxLazySmp(NegamaxSp):
```

Comment:
Sorry to be a pain - could this be moved to a new file?

Reply:
Done and dusted - agree this is clearer